firewood/src/manager.rs

```diff
@@ -339,14 +353,29 @@ impl RevisionManager {
     }
 
     pub fn revision(&self, root_hash: HashKey) -> Result<CommittedRevision, RevisionManagerError> {
-        self.by_hash
+        if let Some(revision) = self
+            .by_hash
             .read()
             .expect("poisoned lock")
             .get(&root_hash)
             .cloned()
+        {
+            return Ok(revision);
+        }
+
+        let root_address = self
+            .root_store
+            .get(&root_hash)
+            .map_err(RevisionManagerError::RootStoreError)?
             .ok_or(RevisionManagerError::RevisionNotFound {
-                provided: root_hash,
-            })
+                provided: root_hash.clone(),
+            })?;
+
+        Ok(Arc::new(NodeStore::with_root(
```

Comment:
If you have 2 requests for the same revision, they will not share the NodeStore, and instead each one will get a separate nodestore with a reference count of 1.

One way to address this is to keep a cache of existing created NodeStores by hash. This might be a use case for weak arcs so those NodeStores will be dropped when unreferenced.